source/server/listener_manager_impl.cc

```diff
@@ -168,6 +168,12 @@ ListenerImpl::ListenerImpl(const envoy::api::v2::Listener& config, const std::st
       listener_tag_(parent_.factory_.nextListenerTag()), name_(name), modifiable_(modifiable),
       workers_started_(workers_started), hash_(hash),
       dynamic_init_manager_(fmt::format("Listener {}", name)),
+      init_watcher_("ListenerImpl",
+                    [this]() -> void {
+                      if (!initialize_canceled_) {
```

Comment:
That's a good question. There's a comment in `ListenerImpl`'s dtor that makes me think it's still necessary:
```c++
// The filter factories may have pending initialize actions (like in the case of RDS). Those
// actions will fire in the destructor to avoid blocking initial server startup. If we are using
// a local init manager we should block the notification from trying to move us from warming to
// active.
```
I don't know enough about this code, but it looks like the `listener_filter_factories_` might contain some initialization targets (notably `RdsRouteConfigSubscription`s) that will signal `ready` on destruction, causing the listener to notify its parent even though it'll be destroyed immediately after.

Reply:
One thing you could potentially do to avoid the extra variable is to allocate `init_watcher_` on the heap, and then explicitly `reset()` it in the destructor. This I think would have the effect of dropping any future completions. WDYT? Not a big deal either way but seems a better use of the new design.